splink/analyse_blocking.py

```diff
@@ -139,3 +139,103 @@ def cumulative_comparisons_generated_by_blocking_rules(
     linker._analyse_blocking_mode = False
 
     return br_comparisons
+
+
+def count_comparisons_from_blocking_rule_pre_filter_conditions_sqls(
+    linker: "Linker", blocking_rule: Union[str, "BlockingRule"]
+):
+    if isinstance(blocking_rule, str):
+        blocking_rule = BlockingRule(blocking_rule, sqlglot_dialect=linker._sql_dialect)
+
+    join_conditions = blocking_rule._join_conditions
+
+    l_cols_sel = []
+    r_cols_sel = []
+    l_cols_gb = []
+    r_cols_gb = []
+    using = []
+    for (
+        i,
+        (l_key, r_key),
+    ) in enumerate(join_conditions):
+        l_cols_sel.append(f"{l_key} as key_{i}")
+        r_cols_sel.append(f"{r_key} as key_{i}")
+        l_cols_gb.append(l_key)
+        r_cols_gb.append(r_key)
+        using.append(f"key_{i}")
+
+    l_cols_sel = ", ".join(l_cols_sel)
+    r_cols_sel = ", ".join(r_cols_sel)
+    l_cols_gb = ", ".join(l_cols_gb)
+    r_cols_gb = ", ".join(r_cols_gb)
+    using = ", ".join(using)
+
+    sqls = []
+
+    if linker._two_dataset_link_only:
+        #    Can just use the raw input datasets
+        keys = list(linker._input_tables_dict.keys())
+        input_tablename_l = linker._input_tables_dict[keys[0]].physical_name
+        input_tablename_r = linker._input_tables_dict[keys[1]].physical_name
+
+    else:
+        input_tablename_l = "__splink__df_concat"
+        input_tablename_r = "__splink__df_concat"
+
+    if not join_conditions:
+        if linker._two_dataset_link_only:
+            sql = f"""
+            SELECT
+                (SELECT COUNT(*) FROM {input_tablename_l})
+                *
+                (SELECT COUNT(*) FROM {input_tablename_r})
+                    AS count_of_pairwise_comparisons_generated
+            """
+        else:
+            sql = """
+            select count(*) * count(*) as count_of_pairwise_comparisons_generated
+            from __splink__df_concat
+
+            """
```

Comment:
This appears to be broken for "stacked" dataframes at present - i.e. a link job where we input a single dataframe with a column detailing which records belongs to which dataset.

```
from splink.duckdb.linker import DuckDBLinker
import pandas as pd
df = pd.read_csv("./tests/datasets/fake_1000_from_splink_demos.csv")
df['source_dataset'] = pd.Series(['a'] * 500 + ['b'] * 500, name='source_dataset')

linker = DuckDBLinker(
    df,
    settings_dict = {
    "link_type": "link_only",
    }
)

linker._initialise_df_concat(True).as_pandas_dataframe()
linker._two_dataset_link_only  # False

br = "levenshtein(first_name, 3)"
display(
    linker._count_num_comparisons_from_blocking_rule_pre_filter_conditions(br)
) # expecting 25k
```

Reply:
Ouch, good spot

Reply:
I think this is broken more widely.  Consider running the following on `master`:
```
from splink.duckdb.linker import DuckDBLinker
import pandas as pd
df = pd.read_csv("./tests/datasets/fake_1000_from_splink_demos.csv")
df['source_dataset'] = pd.Series(['a'] * 500 + ['b'] * 500, name='source_dataset')

linker = DuckDBLinker(
    df,
    settings_dict = {
    "link_type": "link_and_dedupe"
    }
)

import logging 
logging.getLogger("splink").setLevel(1)
linker._initialise_df_concat_with_tf(True).as_pandas_dataframe()
```

You still end up with a column called `__splink_source_dataset`

Reply:
Oh wait - I don't think this is actually broken.    It's a little difficult to explain, but let me try:

- `link_only` works in the general _n_ dataset case, including `n==2`
- In the general case, a self-join of the concatenated datasets is used to generate comparisons, filtering out records where `source_dataset_l = source_dataset_r`
- There is a specific optimisation where splink observes two input datasets that allows an inner join of the raw (non concatenated) input datasets
- This optimisation is not used if you pass Splink a single dataframe that implicitly has two source datasets
- `1,000,000` is therefore the correct result from your code example

Reply:
Separately, there is something a bit weird here for which I will raise a different issue.  Consider running the following on `master`:
```
from splink.duckdb.linker import DuckDBLinker
import pandas as pd
df = pd.read_csv("./tests/datasets/fake_1000_from_splink_demos.csv")
df['source_dataset'] = pd.Series(['a'] * 500 + ['b'] * 500, name='source_dataset')

linker = DuckDBLinker(
    df,
    settings_dict = {
    "link_type": "link_and_dedupe"
    }
)

import logging 
logging.getLogger("splink").setLevel(1)
linker._initialise_df_concat_with_tf(True).as_pandas_dataframe()
```

You still end up with a column called `__splink_source_dataset`, which seems unnecessary

I will raise an issue - I think this probably needs to be fixed separately

Reply:
Yeah, it's definitely unnecessary.

The fix that I put in place was merely intended as a bandage before a more permanent solution was implemented.

Apologies, I should've made an issue at the time.